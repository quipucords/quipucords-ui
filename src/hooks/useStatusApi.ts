import { useCallback } from 'react';
import axios, { type AxiosError, type AxiosResponse, isAxiosError } from 'axios';
import helpers from '../helpers';

type ApiStatusSuccessType = {
  server_version: string;
};

type ApiStatusErrorType = {
  detail?: string;
  message: string;
};

/**
 * A status API call
 */
const useStatusApi = () => {
  const apiCall = useCallback(
    (): Promise<AxiosResponse<ApiStatusSuccessType>> => axios.get(`${process.env.REACT_APP_STATUS_SERVICE}`),
    []
  );

  const callbackSuccess = useCallback((response: AxiosResponse<ApiStatusSuccessType>) => response?.data, []);

  const callbackError = useCallback((error: AxiosError<ApiStatusErrorType>) => {
    return Promise.reject(error);
  }, []);

  const getStatus = useCallback(async () => {
    let response;
    try {
      response = await apiCall();
    } catch (error) {
      if (isAxiosError(error)) {
        return callbackError(error);
      }
      if (!helpers.TEST_MODE) {
        console.error(error);
      }
    }
    return callbackSuccess(response);
  }, [apiCall, callbackSuccess, callbackError]);

  return {
    apiCall,
    callbackError,
    callbackSuccess,
    getStatus
  };
};

export { useStatusApi, type ApiStatusSuccessType };
