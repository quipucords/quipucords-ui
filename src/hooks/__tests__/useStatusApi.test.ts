import { renderHook } from '@testing-library/react';
import axios from 'axios';
import { useStatusApi } from '../useStatusApi';

describe('useStatusApi', () => {
  let hookResult;

  beforeEach(() => {
    const hook = renderHook(() => useStatusApi());
    hookResult = hook?.result?.current;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should attempt an api call to get a status', () => {
    const { apiCall } = hookResult;
    const spyAxios = jest.spyOn(axios, 'get');

    apiCall();
    expect(spyAxios.mock.calls).toMatchSnapshot('apiCall');
  });

  it('should handle success while attempting to get a status', async () => {
    const { getStatus } = hookResult;
    jest.spyOn(axios, 'get').mockImplementation(() => Promise.resolve({}));

    await expect(getStatus()).resolves.toMatchSnapshot('getStatus, success');
  });

  it('should handle errors while attempting to get a status', async () => {
    const { getStatus } = hookResult;
    jest.spyOn(axios, 'get').mockImplementation(() => Promise.reject({ isAxiosError: true, message: 'Mock error' }));

    await expect(getStatus()).rejects.toMatchSnapshot('getStatus, error');
  });

  it('should process an API success response', async () => {
    const { callbackSuccess } = hookResult;

    expect(callbackSuccess({ data: { api_version: 'lorem ipsum', server_version: 'dolor sit' } })).toMatchSnapshot(
      'callbackSuccess'
    );
  });

  it('should process an API error response', async () => {
    const { callbackError } = hookResult;

    await expect(
      callbackError({
        response: {
          data: {
            message: 'Dolor sit'
          }
        }
      })
    ).rejects.toMatchSnapshot('callbackError');
  });
});
