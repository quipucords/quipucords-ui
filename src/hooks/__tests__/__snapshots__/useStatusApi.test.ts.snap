// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStatusApi should attempt an api call to get a status: apiCall 1`] = `
[
  [
    "/api/v1/status/",
  ],
]
`;

exports[`useStatusApi should handle errors while attempting to get a status: getStatus, error 1`] = `
{
  "isAxiosError": true,
  "message": "Mock error",
}
`;

exports[`useStatusApi should handle success while attempting to get a status: getStatus, success 1`] = `undefined`;

exports[`useStatusApi should process an API error response: callbackError 1`] = `
{
  "response": {
    "data": {
      "message": "Dolor sit",
    },
  },
}
`;

exports[`useStatusApi should process an API success response: callbackSuccess 1`] = `
{
  "api_version": "lorem ipsum",
  "server_version": "dolor sit",
}
`;
